pkg/controller/replicaschedulingpreference/controller.go

```diff
@@ -0,0 +1,751 @@
+
+	for _, clusterName := range clusterNames {
+		obj, exists, err := objectGetter(clusterName, key)
+		if err != nil {
+			return nil, nil, err
+		}
+		if !exists {
+			continue
+		}
+		replicas := reflect.ValueOf(obj).Elem().FieldByName("Spec").FieldByName("Replicas").Elem().Int()
+		readyReplicas := reflect.ValueOf(obj).Elem().FieldByName("Status").FieldByName("ReadyReplicas").Int()
+		if replicas == readyReplicas {
+			currentReplicasPerCluster[clusterName] = readyReplicas
+		} else {
+			pods, err := podsGetter(clusterName, obj.(pkgruntime.Object))
+			if err != nil {
+				return nil, nil, err
+			}
+			podStatus := podanalyzer.AnalyzePods(pods, time.Now())
+			currentReplicasPerCluster[clusterName] = int64(podStatus.RunningAndReady) // include pending as well?
+			unschedulable := int64(podStatus.Unschedulable)
+			if unschedulable > 0 {
+				estimatedCapacity[clusterName] = replicas - unschedulable
+			}
+		}
+	}
+	return currentReplicasPerCluster, estimatedCapacity, nil
+}
+
+func updateFederationTargets(fedClient fedclientset.Interface, qualifiedName util.QualifiedName, result map[string]int64) error {
+	newClusterNames := []string{}
+	for name := range result {
+		newClusterNames = append(newClusterNames, name)
+	}
+
+	err := updatePlacement(fedClient, qualifiedName, newClusterNames)
+	if err != nil {
+		return err
+	}
+
+	err = updateOverrides(fedClient, qualifiedName, result)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func updatePlacement(fedClient fedclientset.Interface, qualifiedName util.QualifiedName, newClusterNames []string) error {
+	placement, err := fedClient.FederationV1alpha1().FederatedDeploymentPlacements(qualifiedName.Namespace).Get(qualifiedName.Name, metav1.GetOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		newPlacement := &fedv1a1.FederatedDeploymentPlacement{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: qualifiedName.Namespace,
+				Name:      qualifiedName.Name,
+			},
+			Spec: fedv1a1.FederatedDeploymentPlacementSpec{
+				ClusterNames: newClusterNames,
+			},
+		}
+		_, err := fedClient.FederationV1alpha1().FederatedDeploymentPlacements(qualifiedName.Namespace).Create(newPlacement)
+		return err
+	}
+
+	if placementUpdateNeeded(placement.Spec.ClusterNames, newClusterNames) {
+		// TODO: do we need to make a copy of this object
+		newPlacement := placement
+		newPlacement.Spec.ClusterNames = newClusterNames
+		_, err := fedClient.FederationV1alpha1().FederatedDeploymentPlacements(qualifiedName.Namespace).Update(newPlacement)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func updateOverrides(fedClient fedclientset.Interface, qualifiedName util.QualifiedName, result map[string]int64) error {
+	override, err := fedClient.FederationV1alpha1().FederatedDeploymentOverrides(qualifiedName.Namespace).Get(qualifiedName.Name, metav1.GetOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+
+		// TODO: remove after review and
```

Comment:
I don't understand this comment in its current form.  Please clarify or remove.

Reply:
Removed!